Add missing code to reset RequestEventDispatchers when used in a pool

gavin-norman-sociomantic opened the pull request.

Commit:
Add reset method to RequestEventDispatcher

When storing instances of this struct in a pool, the array fields need to
be cleared when an instance is acquired from the pool. The reset method
encapsulates the required behaviour.

gavin-norman-sociomantic commented:
I was originally viewing this as a bug fix for v4.1.x, so didn't include release notes. Then I thought it's more like a feature PR. What do people think?

nemanja-boric-sociomantic commented:
I think this should go as a bugfix. Otherwise, no real implementation is able to use the facility, which is included in v4.1.x, and that was not our intention.

gavin-norman-sociomantic commented:
Yeah that was my initial thought, until I noticed that it's actually just pure new features. I'll rebase it.

Commit:
Add AcquiredSingleton.acquire override including reset delegate

With the existing, simple acquire() method, there's no way of telling
whether the instance returned was freshly acquired or whether it was
acquired by a previous call. When the singleton instance is acquired from a
pool, we need to be able to reset it only when it's first acquired, not
every time the user accesses it.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -235,6 +235,24 @@ public struct RequestEventDispatcher
     /***************************************************************************
 
+        Resets this instance to its initial state. This method should be called
+        when retriving an instance from a pool or free list.
```
typo `retriving`

gavin-norman-sociomantic replied:
Updated.
